File: src/ml/ann_mlp.cpp
/*
    Created by Abdelaziz Mahdy.
    Licensed: Apache 2.0 license. Copyright (c) 2024 Abdelaziz Mahdy.
*/

#include "ann_mlp.h"
#include "opencv2/ml.hpp"

// ANN_MLP
CvStatus ANN_MLP_Create(PtrANN_MLP *rval)
{
  BEGIN_WRAP
  auto p = cv::ml::ANN_MLP::create();
  rval->ptr = new cv::Ptr<cv::ml::ANN_MLP>(p);
  END_WRAP
}

void ANN_MLP_Close(PtrANN_MLP *self)
{
  self->ptr = nullptr;
  CVD_FREE(self)
}

CvStatus ANN_MLP_SetTrainMethod(PtrANN_MLP self, int method, double param1, double param2)
{
  BEGIN_WRAP
  (*self.ptr)->setTrainMethod(method, param1, param2);
  END_WRAP
}

CvStatus ANN_MLP_GetTrainMethod(PtrANN_MLP self, int *rval)
{
  BEGIN_WRAP
  *rval = (*self.ptr)->getTrainMethod();
  END_WRAP
}

CvStatus ANN_MLP_SetActivationFunction(PtrANN_MLP self, int type, double param1, double param2)
{
  BEGIN_WRAP
  (*self.ptr)->setActivationFunction(type, param1, param2);
  END_WRAP
}

CvStatus ANN_MLP_SetLayerSizes(PtrANN_MLP self, Mat _layer_sizes)
{
  BEGIN_WRAP
  (*self.ptr)->setLayerSizes(*_layer_sizes.ptr);
  END_WRAP
}

CvStatus ANN_MLP_GetLayerSizes(PtrANN_MLP self, Mat *rval)
{
  BEGIN_WRAP
  *rval = {new cv::Mat((*self.ptr)->getLayerSizes())};
  END_WRAP
}

CvStatus ANN_MLP_SetTermCriteria(PtrANN_MLP self, TermCriteria val)
{
  BEGIN_WRAP
  auto tc = cv::TermCriteria(val.type, val.maxCount, val.epsilon);
  (*self.ptr)->setTermCriteria(tc);
  END_WRAP
}

CvStatus ANN_MLP_GetTermCriteria(PtrANN_MLP self, TermCriteria *rval)
{
  BEGIN_WRAP
  auto tc = (*self.ptr)->getTermCriteria();
  *rval = {tc.type, tc.maxCount, tc.epsilon};
  END_WRAP
}

CvStatus ANN_MLP_SetBackpropWeightScale(PtrANN_MLP self, double val)
{
  BEGIN_WRAP
  (*self.ptr)->setBackpropWeightScale(val);
  END_WRAP
}

CvStatus ANN_MLP_GetBackpropWeightScale(PtrANN_MLP self, double *rval)
{
  BEGIN_WRAP
  *rval = (*self.ptr)->getBackpropWeightScale();
  END_WRAP
}

CvStatus ANN_MLP_SetBackpropMomentumScale(PtrANN_MLP self, double val)
{
  BEGIN_WRAP
  (*self.ptr)->setBackpropMomentumScale(val);
  END_WRAP
}

CvStatus ANN_MLP_GetBackpropMomentumScale(PtrANN_MLP self, double *rval)
{
  BEGIN_WRAP
  *rval = (*self.ptr)->getBackpropMomentumScale();
  END_WRAP
}

CvStatus ANN_MLP_SetRpropDW0(PtrANN_MLP self, double val)
{
  BEGIN_WRAP
  (*self.ptr)->setRpropDW0(val);
  END_WRAP
}

CvStatus ANN_MLP_GetRpropDW0(PtrANN_MLP self, double *rval)
{
  BEGIN_WRAP
  *rval = (*self.ptr)->getRpropDW0();
  END_WRAP
}

CvStatus ANN_MLP_SetRpropDWPlus(PtrANN_MLP self, double val)
{
  BEGIN_WRAP
  (*self.ptr)->setRpropDWPlus(val);
  END_WRAP
}

CvStatus ANN_MLP_GetRpropDWPlus(PtrANN_MLP self, double *rval)
{
  BEGIN_WRAP
  *rval = (*self.ptr)->getRpropDWPlus();
  END_WRAP
}

CvStatus ANN_MLP_SetRpropDWMinus(PtrANN_MLP self, double val)
{
  BEGIN_WRAP
  (*self.ptr)->setRpropDWMinus(val);
  END_WRAP
}

CvStatus ANN_MLP_GetRpropDWMinus(PtrANN_MLP self, double *rval)
{
  BEGIN_WRAP
  *rval = (*self.ptr)->getRpropDWMinus();
  END_WRAP
}

CvStatus ANN_MLP_SetRpropDWMin(PtrANN_MLP self, double val)
{
  BEGIN_WRAP
  (*self.ptr)->setRpropDWMin(val);
  END_WRAP
}

CvStatus ANN_MLP_GetRpropDWMin(PtrANN_MLP self, double *rval)
{
  BEGIN_WRAP
  *rval = (*self.ptr)->getRpropDWMin();
  END_WRAP
}

CvStatus ANN_MLP_SetRpropDWMax(PtrANN_MLP self, double val)
{
  BEGIN_WRAP
  (*self.ptr)->setRpropDWMax(val);
  END_WRAP
}

CvStatus ANN_MLP_GetRpropDWMax(PtrANN_MLP self, double *rval)
{
  BEGIN_WRAP
  *rval = (*self.ptr)->getRpropDWMax();
  END_WRAP
}

CvStatus ANN_MLP_SetAnnealInitialT(PtrANN_MLP self, double val)
{
  BEGIN_WRAP
  (*self.ptr)->setAnnealInitialT(val);
  END_WRAP
}

CvStatus ANN_MLP_GetAnnealInitialT(PtrANN_MLP self, double *rval)
{
  BEGIN_WRAP
  *rval = (*self.ptr)->getAnnealInitialT();
  END_WRAP
}

CvStatus ANN_MLP_SetAnnealFinalT(PtrANN_MLP self, double val)
{
  BEGIN_WRAP
  (*self.ptr)->setAnnealFinalT(val);
  END_WRAP
}

CvStatus ANN_MLP_GetAnnealFinalT(PtrANN_MLP self, double *rval)
{
  BEGIN_WRAP
  *rval = (*self.ptr)->getAnnealFinalT();
  END_WRAP
}

CvStatus ANN_MLP_SetAnnealCoolingRatio(PtrANN_MLP self, double val)
{
  BEGIN_WRAP
  (*self.ptr)->setAnnealCoolingRatio(val);
  END_WRAP
}

CvStatus ANN_MLP_GetAnnealCoolingRatio(PtrANN_MLP self, double *rval)
{
  BEGIN_WRAP
  *rval = (*self.ptr)->getAnnealCoolingRatio();
  END_WRAP
}

CvStatus ANN_MLP_SetAnnealItePerStep(PtrANN_MLP self, int val)
{
  BEGIN_WRAP
  (*self.ptr)->setAnnealItePerStep(val);
  END_WRAP
}

CvStatus ANN_MLP_GetAnnealItePerStep(PtrANN_MLP self, int *rval)
{
  BEGIN_WRAP
  *rval = (*self.ptr)->getAnnealItePerStep();
  END_WRAP
}

CvStatus ANN_MLP_Predict(PtrANN_MLP self, Mat samples, Mat results, int flags, float *rval)
{
  BEGIN_WRAP
  *rval = (*self.ptr)->predict(*samples.ptr, *results.ptr, flags);
  END_WRAP
}

CvStatus ANN_MLP_Train(PtrANN_MLP self, PtrTrainData trainData, int flags, bool *rval)
{
  BEGIN_WRAP
  *rval = (*self.ptr)->train(*trainData.ptr, flags);
  END_WRAP
}

CvStatus ANN_MLP_Train_1(PtrANN_MLP self, Mat samples, int layout, Mat responses, bool *rval)
{
  BEGIN_WRAP
  *rval = (*self.ptr)->train(*samples.ptr, layout, *responses.ptr);
  END_WRAP
}

CvStatus ANN_MLP_Clear(PtrANN_MLP self)
{
  BEGIN_WRAP
  (*self.ptr)->clear();
  END_WRAP
}

CvStatus ANN_MLP_Save(PtrANN_MLP self, char *filename)
{
  BEGIN_WRAP
  (*self.ptr)->save(filename);
  END_WRAP
}

CvStatus ANN_MLP_Load(char *filepath, PtrANN_MLP *rval)
{
  BEGIN_WRAP
  cv::Ptr<cv::ml::ANN_MLP> p = cv::ml::ANN_MLP::load(filepath);
  (*rval->ptr) = p;
  END_WRAP
}

CvStatus ANN_MLP_LoadFromString( const char *strModel, const char *objname, PtrANN_MLP *rval)
{
  BEGIN_WRAP
  (*rval->ptr) = cv::Algorithm::loadFromString<cv::ml::ANN_MLP>(strModel, objname);
  END_WRAP
}

File: lib/src/ml/ann_mlp.dart
// ignore_for_file: constant_identifier_names, camel_case_types

import 'dart:ffi' as ffi;
import 'package:ffi/ffi.dart';

import '../core/base.dart';
import '../core/mat.dart';
import '../core/termcriteria.dart';
import '../opencv.g.dart' as cvg;

class ANN_MLP extends CvStruct<cvg.PtrANN_MLP> {
  ANN_MLP._(cvg.PtrANN_MLPPtr ptr) : super.fromPointer(ptr) {
    finalizer.attach(this, ptr.cast());
  }

  factory ANN_MLP.create() {
    final p = calloc<cvg.PtrANN_MLP>();
    cvRun(() => CFFI.ANN_MLP_Create(p));
    return ANN_MLP._(p);
  }

  factory ANN_MLP.load(String filepath) {
    return using<ANN_MLP>((arena) {
      final fp = filepath.toNativeUtf8(allocator: arena).cast<ffi.Char>();
      final p = calloc<cvg.PtrANN_MLP>();
      cvRun(() => CFFI.ANN_MLP_Load(fp, p));
      return ANN_MLP._(p);
    });
  }

  factory ANN_MLP.loadFromString(String strModel, String objname) {
    return using<ANN_MLP>((arena) {
      final sm = strModel.toNativeUtf8(allocator: arena).cast<ffi.Char>();
      final on = objname.toNativeUtf8(allocator: arena).cast<ffi.Char>();
      final p = calloc<cvg.PtrANN_MLP>();
      cvRun(() => CFFI.ANN_MLP_LoadFromString(sm, on, p));
      return ANN_MLP._(p);
    });
  }

  static final finalizer = OcvFinalizer<cvg.PtrANN_MLPPtr>(CFFI.addresses.ANN_MLP_Close);

  int getTrainMethod() {
    return using<int>((arena) {
      final p = arena<ffi.Int>();
      cvRun(() => CFFI.ANN_MLP_GetTrainMethod(ref, p));
      return p.value;
    });
  }

  /// Sets training method and common parameters.
  ///
  /// [method] Default value is [TRAIN_METHODS_RPROP].
  ///
  /// https://docs.opencv.org/4.x/d0/dce/classcv_1_1ml_1_1ANN__MLP.html#a4be093cfd2e743ee2f41e34e50cf3a54
  void setTrainMethod(int method, [double param1 = 0, double param2 = 0]) {
    return using<void>((arena) {
      cvRun(() => CFFI.ANN_MLP_SetTrainMethod(ref, method, param1, param2));
    });
  }

  void setActivationFunction(int type, double param1, double param2) {
    return using<void>((arena) {
      cvRun(() => CFFI.ANN_MLP_SetActivationFunction(ref, type, param1, param2));
    });
  }

  void setLayerSizes(Mat layerSizes) {
    return using<void>((arena) {
      cvRun(() => CFFI.ANN_MLP_SetLayerSizes(ref, layerSizes.ref));
    });
  }

  Mat getLayerSizes() {
    return using<Mat>((arena) {
      final mat = Mat.empty();
      cvRun(() => CFFI.ANN_MLP_GetLayerSizes(ref, mat.ptr));
      return mat;
    });
  }

  void setTermCriteria(TermCriteria val) {
    return using<void>((arena) {
      cvRun(() => CFFI.ANN_MLP_SetTermCriteria(ref, val.toNativePtr(arena).ref));
    });
  }

  TermCriteria getTermCriteria() {
    return using<TermCriteria>((arena) {
      final p = arena<cvg.TermCriteria>();
      cvRun(() => CFFI.ANN_MLP_GetTermCriteria(ref, p));
      return p.ref.toDart();
    });
  }

  void setBackpropWeightScale(double val) {
    return using<void>((arena) {
      cvRun(() => CFFI.ANN_MLP_SetBackpropWeightScale(ref, val));
    });
  }

  double getBackpropWeightScale() {
    return using<double>((arena) {
      final p = arena<ffi.Double>();
      cvRun(() => CFFI.ANN_MLP_GetBackpropWeightScale(ref, p));
      return p.value;
    });
  }

  void setBackpropMomentumScale(double val) {
    return using<void>((arena) {
      cvRun(() => CFFI.ANN_MLP_SetBackpropMomentumScale(ref, val));
    });
  }

  double getBackpropMomentumScale() {
    return using<double>((arena) {
      final p = arena<ffi.Double>();
      cvRun(() => CFFI.ANN_MLP_GetBackpropMomentumScale(ref, p));
      return p.value;
    });
  }

  void setRpropDW0(double val) {
    return using<void>((arena) {
      cvRun(() => CFFI.ANN_MLP_SetRpropDW0(ref, val));
    });
  }

  double getRpropDW0() {
    return using<double>((arena) {
      final p = arena<ffi.Double>();
      cvRun(() => CFFI.ANN_MLP_GetRpropDW0(ref, p));
      return p.value;
    });
  }

  void setRpropDWPlus(double val) {
    return using<void>((arena) {
      cvRun(() => CFFI.ANN_MLP_SetRpropDWPlus(ref, val));
    });
  }

  double getRpropDWPlus() {
    return using<double>((arena) {
      final p = arena<ffi.Double>();
      cvRun(() => CFFI.ANN_MLP_GetRpropDWPlus(ref, p));
      return p.value;
    });
  }

  void setRpropDWMinus(double val) {
    return using<void>((arena) {
      cvRun(() => CFFI.ANN_MLP_SetRpropDWMinus(ref, val));
    });
  }

  double getRpropDWMinus() {
    return using<double>((arena) {
      final p = arena<ffi.Double>();
      cvRun(() => CFFI.ANN_MLP_GetRpropDWMinus(ref, p));
      return p.value;
    });
  }

  void setRpropDWMin(double val) {
    return using<void>((arena) {
      cvRun(() => CFFI.ANN_MLP_SetRpropDWMin(ref, val));
    });
  }

  double getRpropDWMin() {
    return using<double>((arena) {
      final p = arena<ffi.Double>();
      cvRun(() => CFFI.ANN_MLP_GetRpropDWMin(ref, p));
      return p.value;
    });
  }

  void setRpropDWMax(double val) {
    return using<void>((arena) {
      cvRun(() => CFFI.ANN_MLP_SetRpropDWMax(ref, val));
    });
  }

  double getRpropDWMax() {
    return using<double>((arena) {
      final p = arena<ffi.Double>();
      cvRun(() => CFFI.ANN_MLP_GetRpropDWMax(ref, p));
      return p.value;
    });
  }

  void setAnnealInitialT(double val) {
    return using<void>((arena) {
      cvRun(() => CFFI.ANN_MLP_SetAnnealInitialT(ref, val));
    });
  }

  double getAnnealInitialT() {
    return using<double>((arena) {
      final p = arena<ffi.Double>();
      cvRun(() => CFFI.ANN_MLP_GetAnnealInitialT(ref, p));
      return p.value;
    });
  }

  void setAnnealFinalT(double val) {
    return using<void>((arena) {
      cvRun(() => CFFI.ANN_MLP_SetAnnealFinalT(ref, val));
    });
  }

  double getAnnealFinalT() {
    return using<double>((arena) {
      final p = arena<ffi.Double>();
      cvRun(() => CFFI.ANN_MLP_GetAnnealFinalT(ref, p));
      return p.value;
    });
  }

  void setAnnealCoolingRatio(double val) {
    return using<void>((arena) {
      cvRun(() => CFFI.ANN_MLP_SetAnnealCoolingRatio(ref, val));
    });
  }

  double getAnnealCoolingRatio() {
    return using<double>((arena) {
      final p = arena<ffi.Double>();
      cvRun(() => CFFI.ANN_MLP_GetAnnealCoolingRatio(ref, p));
      return p.value;
    });
  }

  void setAnnealItePerStep(int val) {
    return using<void>((arena) {
      cvRun(() => CFFI.ANN_MLP_SetAnnealItePerStep(ref, val));
    });
  }

  int getAnnealItePerStep() {
    return using<int>((arena) {
      final p = arena<ffi.Int>();
      cvRun(() => CFFI.ANN_MLP_GetAnnealItePerStep(ref, p));
      return p.value;
    });
  }

  double predict(Mat samples, Mat results, int flags) {
    return using<double>((arena) {
      final p = arena<ffi.Float>();
      cvRun(() => CFFI.ANN_MLP_Predict(ref, samples.ref, results.ref, flags, p));
      return p.value;
    });
  }

  bool train(cvg.PtrTrainData trainData, int flags) {
    return using<bool>((arena) {
      final p = arena<ffi.Bool>();
      cvRun(() => CFFI.ANN_MLP_Train(ref, trainData, flags, p));
      return p.value;
    });
  }

  bool trainWithSamples(Mat samples, int layout, Mat responses) {
    return using<bool>((arena) {
      final p = arena<ffi.Bool>();
      cvRun(() => CFFI.ANN_MLP_Train_1(ref, samples.ref, layout, responses.ref, p));
      return p.value;
    });
  }

  void clear() {
    return using<void>((arena) {
      cvRun(() => CFFI.ANN_MLP_Clear(ref));
    });
  }

  void save(String filename) {
    return using<void>((arena) {
      final p = filename.toNativeUtf8().cast<ffi.Char>();
      cvRun(() => CFFI.ANN_MLP_Save(ref, p));
      calloc.free(p);
    });
  }

  @override
  List<int> get props => [ptr.address];

  @override
  cvg.PtrANN_MLP get ref => ptr.ref;

  /// The back-propagation algorithm.
  static const int TRAIN_METHODS_BACKPROP = 0;

  /// The RPROP algorithm. See @cite RPROP93 for details.
  static const int TRAIN_METHODS_RPROP = 1;

  /// The simulated annealing algorithm. See @cite Kirkpatrick83 for details.
  static const int TRAIN_METHODS_ANNEAL = 2;

  static const int TRAIN_FLAGS_UPDATE_WEIGHTS = 1;
  static const int TRAIN_FLAGS_NO_INPUT_SCALE = 2;
  static const int TRAIN_FLAGS_NO_OUTPUT_SCALE = 4;

  static const int ACTIVATION_IDENTITY = 0;
  static const int ACTIVATION_SIGMOID_SYM = 1;
  static const int ACTIVATION_GAUSSIAN = 2;
  static const int ACTIVATION_RELU = 3;
  static const int ACTIVATION_LEAKYRELU = 4;
}
